### PR TITLE
Use Long to back the UInt64 type

### PR DESCRIPTION
**Pull request by araspitzu**
The java `Long` uses 64 bits to store a signed integer, since we want to store 64 bits unsigned integers we have a `UInt64` type backed by `BigInt`. This PR improves this by storing an unsigned 64 bits integer in a `Long` (providing unsigned comparison) and using that to back the `UInt64` type.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 1]
```diff
@@ -16,40 +16,36 @@
...
 
-case class UInt64(private val underlying: BigInt) extends Ordered[UInt64] {
+case class UInt64(private val underlying: Long) extends Ordered[UInt64] {
 
-  require(underlying >= 0, s"uint64 must be positive (actual=$underlying)")
-  require(underlying <= UInt64.MaxValueBigInt, s"uint64 must be < 2^64 -1 (actual=$underlying)")
+  override def compare(o: UInt64): Int = UnsignedLongs.compare(underlying, o.underlying)
 
-  override def compare(o: UInt64): Int = underlying.compare(o.underlying)
+  def toByteVector: ByteVector = ByteVector.fromValidHex(UnsignedLongs.toString(underlying, 16))
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Using a codec would be much more efficient that parsing an hex string

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 2]
```diff
@@ -16,40 +16,36 @@
...
-  val MaxValue = UInt64(MaxValueBigInt)
+  val MaxValue = UInt64(hex"0xffffffffffffffff")
 
-  def apply(bin: ByteVector) = new UInt64(new BigInteger(1, bin.toArray))
-
-  def apply(value: Long) = new UInt64(BigInt(value))
+  def apply(bin: ByteVector): UInt64 = bin.toHex match {
+    case hex if hex.isEmpty => UInt64(0)
+    case hex => new UInt64(UnsignedLongs.parseUnsignedLong(hex, 16))
+  }
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Same here, a codec would be more efficient

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 1 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Is it a problem if `toByteVector` here produces padded byte arrays? This is how the `uint64` codec in `CommonsCodec` behaves.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 2 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Do you mean using the `uint64` scodec? We can't use it here because it's defined using this constructor so it would cause an infinite loop.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 3]
```diff
@@ -16,40 +16,42 @@
...
+  val MaxValue = UInt64(hex"0xffffffffffffffff")
 
-  val MaxValue = UInt64(MaxValueBigInt)
-
-  def apply(bin: ByteVector) = new UInt64(new BigInteger(1, bin.toArray))
-
-  def apply(value: Long) = new UInt64(BigInt(value))
+  def apply(bin: ByteVector): UInt64 = UInt64(bin.toLong(signed = false))
 
   object Conversions {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I think these conversions aren't needed anymore

Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Actually they're still in use by [TlvCodecs](https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/wire/TlvCodecs.scala#L36), we could use a more explicit conversion there but i'm afraid it becomes quite uglier because we'd need to do unsigned comparisons (which is currently achieved by implicitly converting the right hand side to a `UInt64` and using its `compare` implementation)

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Ah no you're right, I completely messed that up :)

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 4]
```diff
@@ -16,40 +16,42 @@
...
-case class UInt64(private val underlying: BigInt) extends Ordered[UInt64] {
+case class UInt64(private val underlying: Long) extends Ordered[UInt64] {
 
-  require(underlying >= 0, s"uint64 must be positive (actual=$underlying)")
-  require(underlying <= UInt64.MaxValueBigInt, s"uint64 must be < 2^64 -1 (actual=$underlying)")
+  override def compare(o: UInt64): Int = UnsignedLongs.compare(underlying, o.underlying)
 
-  override def compare(o: UInt64): Int = underlying.compare(o.underlying)
+  def toByteVector: ByteVector = this match {
+    case x if x <= UInt64(255) => ByteVector.fromLong(underlying, size = 1)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: A more compact way of doing this:
```scala
def toByteVector: ByteVector = underlying match {
  case 0 => hex"00"
  case _ => ByteVector.fromLong(underlying).dropWhile(_ == 0)
}
```

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 5]
```diff
@@ -16,40 +16,42 @@
...
+    case x if x <= UInt64(281474976710655L) => ByteVector.fromLong(underlying, size = 6)
+    case x if x <= UInt64(72057594037927935L) => ByteVector.fromLong(underlying, size = 7)
+    case _ => ByteVector.fromLong(underlying, size = 8)
+  }
 
-  def toByteVector: ByteVector = ByteVector.view(underlying.toByteArray.takeRight(8))
+  def toBigInt: BigInt = BigInt(toString)
 
-  def toBigInt: BigInt = underlying
+  def toLong: Long = underlying
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I would remove this: it's unused and can be confusing when UInt64 overflows a positive Long, don't you think?

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 6]
```diff
@@ -16,40 +16,42 @@
 
 package fr.acinq.eclair
 
-import java.math.BigInteger
-
+import com.google.common.primitives.UnsignedLongs
 import scodec.bits.ByteVector
+import scodec.bits._
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: nit: you only need `scodec.bits.HexStringSyntax`, no need to import everything

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 7]
```diff
@@ -16,40 +16,42 @@
...
 
-case class UInt64(private val underlying: BigInt) extends Ordered[UInt64] {
+case class UInt64(private val underlying: Long) extends Ordered[UInt64] {
 
-  require(underlying >= 0, s"uint64 must be positive (actual=$underlying)")
-  require(underlying <= UInt64.MaxValueBigInt, s"uint64 must be < 2^64 -1 (actual=$underlying)")
+  override def compare(o: UInt64): Int = UnsignedLongs.compare(underlying, o.underlying)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I would move the `compareUnsigned` that's currently in `MilliSatoshi` and put it here: this way the use of `UnsignedLongs` is constrained to this file, which is really the only place where we want to hide it.
What do you think?

File: eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala [hunk 8]
```diff
@@ -31,6 +31,7 @@ class UInt64Spec extends FunSuite {
     assert(b < a)
     assert(z < a && z < b && z < c)
     assert(a == a)
+    assert(a == UInt64.MaxValue)
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala]: I think it's also worth enriching these tests with more edge cases around `Long.MaxValue`.
I'd also test the toString method.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 9]
```diff
@@ -16,40 +16,42 @@
...
+  }
 
-  def toByteVector: ByteVector = ByteVector.view(underlying.toByteArray.takeRight(8))
+  def toBigInt: BigInt = BigInt(toString)
 
-  def toBigInt: BigInt = underlying
+  def toLong: Long = underlying
 
-  override def toString: String = underlying.toString
+  override def toString: String = UnsignedLongs.toString(underlying, 10)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Maybe adding `toHexString` can be useful to debug big values?

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 10]
```diff
@@ -16,40 +16,42 @@
...
+    case x if x <= UInt64(16777215) => ByteVector.fromLong(underlying, size = 3)
+    case x if x <= UInt64(4294967295L) => ByteVector.fromLong(underlying, size = 4)
+    case x if x <= UInt64(1099511627775L) => ByteVector.fromLong(underlying, size = 5)
+    case x if x <= UInt64(281474976710655L) => ByteVector.fromLong(underlying, size = 6)
+    case x if x <= UInt64(72057594037927935L) => ByteVector.fromLong(underlying, size = 7)
+    case _ => ByteVector.fromLong(underlying, size = 8)
+  }
 
-  def toByteVector: ByteVector = ByteVector.view(underlying.toByteArray.takeRight(8))
+  def toBigInt: BigInt = BigInt(toString)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: It feels wasteful to go a string to convert, I'd use `BigInt(toByteVector.toArray)`

File: eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala (hunk 8 above)
Comment (t-bast) [reply eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala]: Overall this `UInt64` is full of subtleties and everything could go wrong if we didn't handle an edge case well, so I'd add a lot more things here to feel safe...
That's also a good candidate for fuzzing when we add that to eclair.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 5 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Agreed, it's confusing and mostly wrong because a signed Long can't accommodate an uint64.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 4 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Nice one! I was a bit wary of doing so many comparisons and this solves it elegantly.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 9 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Not sure about that, using `toByteVector` the caller can easily get the hex string calling `toHex`.

File: eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala (hunk 8 above)
Comment (araspitzu) [reply eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala]: I did some similar testing in [here](https://github.com/ACINQ/eclair/blob/master/eclair-core/src/test/scala/fr/acinq/eclair/CoinUtilsSpec.scala#L32), it's related but i think is worth repeating the tests in `UInt64Spec`. As you said this is plenty of subtleties and thorough testing is the first step to get them right.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 9 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: True

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 7 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I like this proposal, will do a `UInt64.compareUnsigned(other: MilliSatoshi)`

Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: bfd7f6b moves the unsigned comparison into the `UInt64` class, also the comparison now takes into account the possible negative values for MilliSatoshi and guarantees that `UInt64(_) > MilliSatoshi(y)` for any `y` smaller than `0`

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 11]
```diff
@@ -16,40 +16,38 @@
...
 
-case class UInt64(private val underlying: BigInt) extends Ordered[UInt64] {
-
-  require(underlying >= 0, s"uint64 must be positive (actual=$underlying)")
-  require(underlying <= UInt64.MaxValueBigInt, s"uint64 must be < 2^64 -1 (actual=$underlying)")
+case class UInt64(private val underlying: Long) extends Ordered[UInt64] {
 
-  override def compare(o: UInt64): Int = underlying.compare(o.underlying)
+  override def compare(o: UInt64): Int = UnsignedLongs.compare(underlying, o.underlying)
+  def compareUnsigned(other: MilliSatoshi) = other.toLong match {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I would even rename that to `compare` (since it's implicit that we compare unsigned because UInt64 is unsigned).
And I would also define:
```scala
def <(other: MilliSatoshi): Boolean
def >(other: MilliSatoshi): Boolean
def <=(other: MilliSatoshi): Boolean
def >=(other: MilliSatoshi): Boolean
```
like what we do in MilliSatoshi to compare with BtcAmount.
What do you think?

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 12]
```diff
@@ -148,7 +148,7 @@ object Commitments {
 
     val htlcValueInFlight = outgoingHtlcs.map(_.add.amountMsat).sum
     // we must use unsigned comparison here because 'maxHtlcValueInFlightMsat' is effectively an uint64 and can exceed the capacity of MilliSatoshi class
-    if (htlcValueInFlight.compareUnsigned(commitments1.remoteParams.maxHtlcValueInFlightMsat) > 0) {
+    if (commitments1.remoteParams.maxHtlcValueInFlightMsat.compareUnsigned(htlcValueInFlight) < 0) {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: If you implement `<` you can write `commitments1.remoteParams.maxHtlcValueInFlightMsat < htlcValueInFlight` which is much better :)

File: eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala [hunk 13]
```diff
@@ -22,23 +22,52 @@ import scodec.bits._
...
     assert(b.toString === "18446744073709551614")
     assert(c.toByteVector === hex"0x2a")
     assert(c.toString === "42")
     assert(z.toByteVector === hex"0x00")
     assert(z.toString === "0")
+    assert(UInt64(hex"0xff").toByteVector == hex"0xff")
+    assert(UInt64(hex"0x800").toByteVector == hex"0x800")
+  }
+
+  test("use unsigned comparison when comparing millisatoshis to uint64") {
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala]: And those tests would also be much easier to read if it used `<` and `>`!

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 11 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I see the point of your suggestion but i also like having this (unsigned comparison) very explicit, after all is a bit of a special case; defining those comparison operators feels like hiding the complexity because at first glance the reader won't see that a special comparison is used. With that said i'm not strongly opposed and if we want to do it this way i'm okay as well (probably a comment in the code is enough?).

Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I think it's weird that `UInt64` (which is explicitly unsigned) defines a method `compareUnsigned` :)
Especially since it's not really a `compareUnsigned` anymore: we take into account the `MilliSatoshi` sign: if negative it's smaller than our `UInt64`, otherwise we are just comparing two positive values (and the fact that we internally use a compareUnsigned on long is just a side effect of how we encode our UInt64).

Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: That's a fair point, `compareUnsigned` is a bit weird since is defined on an unsigned type. Okay i will do that, let's see how it goes through the review :)

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 14]
```diff
@@ -16,40 +16,43 @@
...
-  override def compare(o: UInt64): Int = underlying.compare(o.underlying)
+  def <(other: MilliSatoshi): Boolean = compare(other) < 0
+  def >(other: MilliSatoshi): Boolean = compare(other) > 0
+  def <=(other: MilliSatoshi): Boolean = compare(other) <= 0
+  def >=(other: MilliSatoshi): Boolean = compare(other) >= 0
 
-  def toByteVector: ByteVector = ByteVector.view(underlying.toByteArray.takeRight(8))
+  def toByteVector: ByteVector = underlying match {
+    case 0 => hex"00"
+    case _ => ByteVector.fromLong(underlying).dropWhile(_ == 0)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: The `dropWhile` shouldn't be necessary anymore

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala [hunk 15]
```diff
@@ -183,8 +182,7 @@ object Commitments {
     val incomingHtlcs = reduced.htlcs.filter(_.direction == IN)
 
     val htlcValueInFlight = incomingHtlcs.map(_.add.amountMsat).sum
-    // we must use unsigned comparison here because 'maxHtlcValueInFlightMsat' is effectively an uint64 and can exceed the capacity of MilliSatoshi class
-    if (htlcValueInFlight.compareUnsigned(commitments1.localParams.maxHtlcValueInFlightMsat) > 0) {
+    if (commitments1.localParams.maxHtlcValueInFlightMsat < htlcValueInFlight) {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala]: The removed comment would probably have helped somewhere around line 30 in UInt64.scala.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 14 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: I think we need it, otherwise we will always have something that's 8-bytes long?

**Comment by pm47** [comment]
Did we keep the multiple `UInt64.toBigInt.toLong` on purpose?

**Comment by t-bast** [comment]
> Did we keep the multiple UInt64.toBigInt.toLong on purpose?

What's the behavior for a UInt64 that overflows Long.MaxValue?
Is `.toBigInt.toLong` different from `underlying`?

**Comment by araspitzu** [comment]
The behavior or `toBigInt.toLong` is equivalent to reading the `underlying` Long of `UInt64`, reference api https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#longValue() . There is also the `longValueExact` function that avoids losing precision but unfortunately is only available for JDK >= 8.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala [hunk 16]
```diff
@@ -16,40 +16,43 @@
...
+  }
 
-  def toBigInt: BigInt = underlying
+  def toBigInt: BigInt = BigInt(signum = 1, toByteVector.toArray)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Feels a bit weird to go through `toByteVector` for converting to `BigInt`. Stackoverflow suggests:

```suggestion
  def toBigInt: BigInt = (BigInt(underlying >>> 1) << 1) + (underlying & 1)
```

source: https://stackoverflow.com/questions/21212993/unsigned-variables-in-scala

Tests pass with this change.

File: eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala [hunk 17]
```diff
@@ -22,23 +22,52 @@ import scodec.bits._
...
+    assert(a == UInt64.MaxValue)
+    assert(l.toByteVector == hex"7fffffffffffffff")
+    assert(l.toString == Long.MaxValue.toString)
+    assert(l.toBigInt == BigInt(Long.MaxValue))
+    assert(l1.toByteVector == hex"8000000000000000")
+    assert(l1.toString == "9223372036854775808")
+    assert(l1.toBigInt == BigInt("9223372036854775808"))
     assert(a.toByteVector === hex"0xffffffffffffffff")
-    assert(a.toString === "18446744073709551615")
+    assert(a.toString === "18446744073709551615") // 2^64 - 1
```
Comment (pm47) [inline eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala]: I feel we are missing a few tests about `toBigInt`. Patch:

```diff
diff --git a/eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala b/eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala
index 16f58f83..cbd75988 100644
--- a/eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala
+++ b/eclair-core/src/test/scala/fr/acinq/eclair/UInt64Spec.scala
@@ -38,20 +38,31 @@ class UInt64Spec extends FunSuite {
     assert(z < a && z < b && z < c && z < l && c < l && l < l1 && l < b && l < a)
     assert(a == a)
     assert(a == UInt64.MaxValue)
+
     assert(l.toByteVector == hex"7fffffffffffffff")
     assert(l.toString == Long.MaxValue.toString)
     assert(l.toBigInt == BigInt(Long.MaxValue))
+
     assert(l1.toByteVector == hex"8000000000000000")
     assert(l1.toString == "9223372036854775808")
     assert(l1.toBigInt == BigInt("9223372036854775808"))
+
     assert(a.toByteVector === hex"0xffffffffffffffff")
     assert(a.toString === "18446744073709551615") // 2^64 - 1
+    assert(a.toBigInt === BigInt("18446744073709551615"))
+
     assert(b.toByteVector === hex"0xfffffffffffffffe")
     assert(b.toString === "18446744073709551614")
+    assert(b.toBigInt === BigInt("18446744073709551614"))
+
     assert(c.toByteVector === hex"0x2a")
     assert(c.toString === "42")
+    assert(c.toBigInt === BigInt("42"))
+
     assert(z.toByteVector === hex"0x00")
     assert(z.toString === "0")
+    assert(z.toBigInt === BigInt("0"))
+
     assert(UInt64(hex"0xff").toByteVector == hex"0xff")
     assert(UInt64(hex"0x800").toByteVector == hex"0x800")
   }

```

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 14 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: Replacing the whole method by:

```scala
  def toByteVector: ByteVector = ByteVector.fromLong(underlying)
```

should work (except that it makes cosmetic tests fail). I think it is simpler that way, do you see a potential problem?

**Comment by araspitzu** [comment]
> Replacing the whole method by:
>
>  def toByteVector: ByteVector = ByteVector.fromLong(underlying)
> should work (except that it makes cosmetic tests fail). I think it is simpler that way, do you see a potential problem?

I don't see problems with that, currently `toByteVector` is used by `CommonsCodecs`, `TlvCodecs` and `toBigInt`. I just tried to remove the truncation and the codec tests are passing, unless in some part of the code we require the `UInt64` to be minimally encoded in a byte vector this change should be fine.

File: eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala (hunk 16 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/UInt64.scala]: The change have been implemented in 8c60d6a, while a bit less readable i think this solution is elegant and more efficient than going through the byte vector conversion.
